logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -447,6 +453,102 @@ public final IRubyObject collectStats(final ThreadContext context) throws IOExce
+
+        final LongCounter eventsFilteredCounter = initOrGetCounterMetric(context, eventsNamespace, MetricKeys.FILTERED_KEY);
+        final FlowMetric filterThroughput = new FlowMetric("filter_throughput", eventsFilteredCounter, uptimeInSeconds);
+        this.flowMetrics.add(filterThroughput);
+        storeMetric(context, flowNamespace, filterThroughput);
+
+        final LongCounter eventsOutCounter = initOrGetCounterMetric(context, eventsNamespace, MetricKeys.OUT_KEY);
+        final FlowMetric outputThroughput = new FlowMetric("output_throughput", eventsOutCounter, uptimeInSeconds);
+        this.flowMetrics.add(outputThroughput);
+        storeMetric(context, flowNamespace, outputThroughput);
+
+        final LongCounter queuePushWaitInMillis = initOrGetCounterMetric(context, eventsNamespace, JRubyWrappedWriteClientExt.PUSH_DURATION_KEY);
+        final FlowMetric backpressureFlow = new FlowMetric("backpressure", queuePushWaitInMillis, uptimeInMillis);
+        this.flowMetrics.add(backpressureFlow);
+        storeMetric(context, flowNamespace, backpressureFlow);
+
+        final LongCounter durationInMillis = initOrGetCounterMetric(context, eventsNamespace, MetricKeys.DURATION_IN_MILLIS_KEY);
+        final FlowMetric concurrencyFlow = new FlowMetric("concurrency", durationInMillis, uptimeInMillis);
+        this.flowMetrics.add(concurrencyFlow);
+        storeMetric(context, flowNamespace, concurrencyFlow);
+
+        return context.nil;
+    }
+
+    @JRubyMethod(name = "collect_flow_metrics")
+    public final IRubyObject collectFlowMetrics(final ThreadContext context) {
+        this.flowMetrics.forEach(FlowMetric::capture);
+        return context.nil;
+    }
+
+    LongCounter initOrGetCounterMetric(final ThreadContext context,
+                                       final RubySymbol[] subPipelineNamespacePath,
+                                       final RubySymbol metricName) {
+        final IRubyObject collector = this.metric.collector(context);
+        final IRubyObject fullNamespace = RubyArray.newArray(context.runtime, fullNamespacePath(subPipelineNamespacePath));
+
+        final IRubyObject retrievedMetric = collector.callMethod(context, "get", new IRubyObject[]{fullNamespace, metricName, context.runtime.newSymbol("counter")});
+        return retrievedMetric.toJava(LongCounter.class);
+    }
+
+    UptimeMetric initOrGetUptimeMetric(final ThreadContext context,
+                                       final RubySymbol[] subPipelineNamespacePath,
+                                       final RubySymbol uptimeMetricName) {
+        final IRubyObject collector = this.metric.collector(context);
+        final IRubyObject fullNamespace = RubyArray.newArray(context.runtime, fullNamespacePath(subPipelineNamespacePath));
+
+        final IRubyObject retrievedMetric = collector.callMethod(context, "get", new IRubyObject[]{fullNamespace, uptimeMetricName, context.runtime.newSymbol("uptime")});
```

Comment:
wooh, for the long term we need to eliminate these back and forth logics...

Reply:
they. are. awful.